In [1]:
import sys
import os
import pandas as pd

# Add the parent directory (project root) to sys.path
sys.path.append(os.path.abspath('..'))

# Now use an absolute import
from src.extraction.statsbomb import fetch_statsbomb_data

In [40]:
# Remove the display limit for columns
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_seq_items', None)

In [3]:
target_matches, events_df, lineups_df = fetch_statsbomb_data()


C:\Users\dell\Data Science\Projects\002_FIFA2022_ETL_RAG_pipeline\fifa2022-etl-rag-pipeline\.venv\Lib\site-packages\statsbombpy\api_client.py:27: NoAuthWarning: credentials were not supplied. open data access only
  warnings.warn(


## Target Matches

In [4]:
target_matches.head() 


,match_id,match_date,kick_off,home_score,away_score,match_status,match_status_360,last_updated,last_updated_360,match_week,competition_id,competition_country_name,competition_name,competition,season_id,season,home_team_id,home_team,home_team_gender,home_team_group,home_team_country_id,home_team_country_name,away_team_id,away_team,away_team_gender,away_team_group,away_team_country_id,away_team_country_name,competition_stage_id,competition_stage,stadium_id,stadium,stadium_country_id,stadium_country_name,referee_id,referee,referee_country_id,referee_country_name,home_managers,away_managers,home_manager_id,home_manager_name,home_manager_nickname,home_manager_dob,home_manager_country_id,home_manager_country_name,away_manager_id,away_manager_name,away_manager_nickname,away_manager_dob,away_manager_country_id,away_manager_country_name,data_version,shot_fidelity_version,xy_fidelity_version
0,3869685,2022-12-18,17:00:00.000,3,3,available,available,2024-12-16T10:15:11.055845,2024-12-16T10:21:13.710934,7,43,International,FIFA World Cup,International - FIFA World Cup,106,2022,779,Argentina,male,NaN,11,Argentina,771,France,male,NaN,78,France,26,Final,1001114,Lusail Stadium,185,Qatar,367.0,Szymon Marciniak,182.0,Poland,Lionel Sebastián Scaloni,Didier Deschamps,5677,Lionel Sebastián Scaloni,NaN,1978-05-16,11,Argentina,545,Didier Deschamps,NaN,1968-10-15,78,France,1.1.0,2,2
1,3869519,2022-12-13,21:00:00.000,3,0,available,available,2023-04-19T06:53:52.197711,2023-04-26T22:32:37.808359,6,43,International,FIFA World Cup,International - FIFA World Cup,106,2022,779,Argentina,male,NaN,11,Argentina,785,Croatia,male,NaN,56,Croatia,15,Semi-finals,1001114,Lusail Stadium,185,Qatar,269.0,Daniele Orsato,112.0,Italy,Lionel Sebastián Scaloni,Zlatko Dalić,5677,Lionel Sebastián Scaloni,NaN,1978-05-16,11,Argentina,307,Zlatko Dalić,NaN,1966-10-26,56,Croatia,1.1.0,2,2
2,3869552,2022-12-14,21:00:00.000,2,0,available,available,2024-02-06T16:36:46.784457,2024-02-06T16:45:42.775801,6,43,International,FIFA World Cup,International - FIFA World Cup,106,2022,771,France,male,NaN,78,France,788,Morocco,male,NaN,154,Morocco,15,Semi-finals,1001113,Al Bayt Stadium,185,Qatar,740.0,César Arturo Ramos Palazuelos,147.0,Mexico,Didier Deschamps,Hoalid Regragui,545,Didier Deschamps,NaN,1968-10-15,78,France,1001086,Hoalid Regragui,NaN,1975-09-23,154,Morocco,1.1.0,2,2


In [5]:
target_matches.info()
target_matches.shape # (3 Rows, 55 Column)

<class 'pandas.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 55 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   match_id                   3 non-null      int64  
 1   match_date                 3 non-null      str    
 2   kick_off                   3 non-null      str    
 3   home_score                 3 non-null      int64  
 4   away_score                 3 non-null      int64  
 5   match_status               3 non-null      str    
 6   match_status_360           3 non-null      str    
 7   last_updated               3 non-null      str    
 8   last_updated_360           3 non-null      str    
 9   match_week                 3 non-null      int64  
 10  competition_id             3 non-null      int64  
 11  competition_country_name   3 non-null      str    
 12  competition_name           3 non-null      str    
 13  competition                3 non-null      str    
 14  season_id

(3, 55)

In [6]:
[target_matches.columns]

[Index(['match_id', 'match_date', 'kick_off', 'home_score', 'away_score',
        'match_status', 'match_status_360', 'last_updated', 'last_updated_360',
        'match_week', 'competition_id', 'competition_country_name',
        'competition_name', 'competition', 'season_id', 'season',
        'home_team_id', 'home_team', 'home_team_gender', 'home_team_group',
        'home_team_country_id', 'home_team_country_name', 'away_team_id',
        'away_team', 'away_team_gender', 'away_team_group',
        'away_team_country_id', 'away_team_country_name',
        'competition_stage_id', 'competition_stage', 'stadium_id', 'stadium',
        'stadium_country_id', 'stadium_country_name', 'referee_id', 'referee',
        'referee_country_id', 'referee_country_name', 'home_managers',
        'away_managers', 'home_manager_id', 'home_manager_name',
        'home_manager_nickname', 'home_manager_dob', 'home_manager_country_id',
        'home_manager_country_name', 'away_manager_id', 'away_manager_n

In [7]:
# null values
target_matches.loc[:, target_matches.isna().any()].columns


Index(['home_team_group', 'away_team_group', 'home_manager_nickname',
       'away_manager_nickname'],
      dtype='str')

In [8]:
columns = ['match_id', 'match_date', 'kick_off', 'home_score', 'away_score',
        'match_status',
        'match_week', 'competition_id', 'competition_country_name',
        'competition_name', 'competition', 'season_id', 'season',
        'home_team_id', 'home_team',
        'home_team_country_id', 'home_team_country_name', 'away_team_id',
        'away_team',
        'away_team_country_id', 'away_team_country_name', 'stadium_id', 'stadium',
        'stadium_country_id', 'stadium_country_name', 'referee_id', 'referee',
        'referee_country_id', 'referee_country_name', 'home_managers',
        'away_managers', 'home_manager_id', 'home_manager_name', 'home_manager_dob', 'home_manager_country_id',
        'home_manager_country_name', 'away_manager_id', 'away_manager_name',
         'away_manager_dob', 'away_manager_country_id',
        'away_manager_country_name', 'data_version', 'shot_fidelity_version',
        'xy_fidelity_version']
target_matches_columns = target_matches[columns]

In [9]:
target_matches_columns.head()


,match_id,match_date,kick_off,home_score,away_score,match_status,match_week,competition_id,competition_country_name,competition_name,competition,season_id,season,home_team_id,home_team,home_team_country_id,home_team_country_name,away_team_id,away_team,away_team_country_id,away_team_country_name,stadium_id,stadium,stadium_country_id,stadium_country_name,referee_id,referee,referee_country_id,referee_country_name,home_managers,away_managers,home_manager_id,home_manager_name,home_manager_dob,home_manager_country_id,home_manager_country_name,away_manager_id,away_manager_name,away_manager_dob,away_manager_country_id,away_manager_country_name,data_version,shot_fidelity_version,xy_fidelity_version
0,3869685,2022-12-18,17:00:00.000,3,3,available,7,43,International,FIFA World Cup,International - FIFA World Cup,106,2022,779,Argentina,11,Argentina,771,France,78,France,1001114,Lusail Stadium,185,Qatar,367.0,Szymon Marciniak,182.0,Poland,Lionel Sebastián Scaloni,Didier Deschamps,5677,Lionel Sebastián Scaloni,1978-05-16,11,Argentina,545,Didier Deschamps,1968-10-15,78,France,1.1.0,2,2
1,3869519,2022-12-13,21:00:00.000,3,0,available,6,43,International,FIFA World Cup,International - FIFA World Cup,106,2022,779,Argentina,11,Argentina,785,Croatia,56,Croatia,1001114,Lusail Stadium,185,Qatar,269.0,Daniele Orsato,112.0,Italy,Lionel Sebastián Scaloni,Zlatko Dalić,5677,Lionel Sebastián Scaloni,1978-05-16,11,Argentina,307,Zlatko Dalić,1966-10-26,56,Croatia,1.1.0,2,2
2,3869552,2022-12-14,21:00:00.000,2,0,available,6,43,International,FIFA World Cup,International - FIFA World Cup,106,2022,771,France,78,France,788,Morocco,154,Morocco,1001113,Al Bayt Stadium,185,Qatar,740.0,César Arturo Ramos Palazuelos,147.0,Mexico,Didier Deschamps,Hoalid Regragui,545,Didier Deschamps,1968-10-15,78,France,1001086,Hoalid Regragui,1975-09-23,154,Morocco,1.1.0,2,2


**Decision** <br>
Separating the ```competition``` details from the match dataset, keeping only the competition ID and refer it later:<br>

table_name: competition<br>

|key|value|
|---|----|
|competition_id | 43|
|competition_name | FIFA World Cup|
|competition_country_name| International|
|season_id | 106|
|season | 2022|

In [10]:
target_matches_columns.drop(columns=['competition_name','season_id','season','competition','competition_country_name'], inplace=True)

In [11]:
target_matches_columns.head()

,match_id,match_date,kick_off,home_score,away_score,match_status,match_week,competition_id,home_team_id,home_team,home_team_country_id,home_team_country_name,away_team_id,away_team,away_team_country_id,away_team_country_name,stadium_id,stadium,stadium_country_id,stadium_country_name,referee_id,referee,referee_country_id,referee_country_name,home_managers,away_managers,home_manager_id,home_manager_name,home_manager_dob,home_manager_country_id,home_manager_country_name,away_manager_id,away_manager_name,away_manager_dob,away_manager_country_id,away_manager_country_name,data_version,shot_fidelity_version,xy_fidelity_version
0,3869685,2022-12-18,17:00:00.000,3,3,available,7,43,779,Argentina,11,Argentina,771,France,78,France,1001114,Lusail Stadium,185,Qatar,367.0,Szymon Marciniak,182.0,Poland,Lionel Sebastián Scaloni,Didier Deschamps,5677,Lionel Sebastián Scaloni,1978-05-16,11,Argentina,545,Didier Deschamps,1968-10-15,78,France,1.1.0,2,2
1,3869519,2022-12-13,21:00:00.000,3,0,available,6,43,779,Argentina,11,Argentina,785,Croatia,56,Croatia,1001114,Lusail Stadium,185,Qatar,269.0,Daniele Orsato,112.0,Italy,Lionel Sebastián Scaloni,Zlatko Dalić,5677,Lionel Sebastián Scaloni,1978-05-16,11,Argentina,307,Zlatko Dalić,1966-10-26,56,Croatia,1.1.0,2,2
2,3869552,2022-12-14,21:00:00.000,2,0,available,6,43,771,France,78,France,788,Morocco,154,Morocco,1001113,Al Bayt Stadium,185,Qatar,740.0,César Arturo Ramos Palazuelos,147.0,Mexico,Didier Deschamps,Hoalid Regragui,545,Didier Deschamps,1968-10-15,78,France,1001086,Hoalid Regragui,1975-09-23,154,Morocco,1.1.0,2,2


## Line Ups

In [12]:
lineups_df.head()

,player_id,player_name,player_nickname,jersey_number,country,cards,positions,match_id,team
0,2972,Marcus Thuram,NaN,26,France,"[{'time': '86:24', 'card_type': 'Yellow Card', 'reason': 'Foul Committed', 'period': 2}]","[{'position_id': 23, 'position': 'Center Forward', 'from': '41:01', 'to': '41:20', 'from_period': 1, 'to_period': 1, 'start_reason': 'Substitution - On (Tactical)', 'end_reason': 'Tactical Shift'}, {'position_id': 21, 'position': 'Left Wing', 'from': '41:20', 'to': '64:09', 'from_period': 1, 'to_period': 2, 'start_reason': 'Tactical Shift', 'end_reason': 'Tactical Shift'}, {'position_id': 23, 'position': 'Center Forward', 'from': '64:09', 'to': '72:52', 'from_period': 2, 'to_period': 2, 'start_reason': 'Tactical Shift', 'end_reason': 'Tactical Shift'}, {'position_id': 22, 'position': 'Right Center Forward', 'from': '72:52', 'to': '74:05', 'from_period': 2, 'to_period': 2, 'start_reason': 'Tactical Shift', 'end_reason': 'Tactical Shift'}, {'position_id': 16, 'position': 'Left Midfield', 'from': '74:05', 'to': None, 'from_period': 2, 'to_period': None, 'start_reason': 'Tactical Shift', 'end_reason': 'Final Whistle'}]",3869685,France
1,3009,Kylian Mbappé Lottin,Kylian Mbappé,10,France,[],"[{'position_id': 21, 'position': 'Left Wing', 'from': '00:00', 'to': '41:20', 'from_period': 1, 'to_period': 1, 'start_reason': 'Starting XI', 'end_reason': 'Tactical Shift'}, {'position_id': 23, 'position': 'Center Forward', 'from': '41:20', 'to': '64:09', 'from_period': 1, 'to_period': 2, 'start_reason': 'Tactical Shift', 'end_reason': 'Tactical Shift'}, {'position_id': 21, 'position': 'Left Wing', 'from': '64:09', 'to': '72:52', 'from_period': 2, 'to_period': 2, 'start_reason': 'Tactical Shift', 'end_reason': 'Tactical Shift'}, {'position_id': 24, 'position': 'Left Center Forward', 'from': '72:52', 'to': None, 'from_period': 2, 'to_period': None, 'start_reason': 'Tactical Shift', 'end_reason': 'Final Whistle'}]",3869685,France
2,3026,Adrien Rabiot,NaN,14,France,"[{'time': '54:12', 'card_type': 'Yellow Card', 'reason': 'Foul Committed', 'period': 2}]","[{'position_id': 11, 'position': 'Left Defensive Midfield', 'from': '00:00', 'to': '95:10', 'from_period': 1, 'to_period': 3, 'start_reason': 'Starting XI', 'end_reason': 'Substitution - Off (Tactical)'}]",3869685,France
3,3099,Hugo Lloris,NaN,1,France,[],"[{'position_id': 1, 'position': 'Goalkeeper', 'from': '00:00', 'to': None, 'from_period': 1, 'to_period': None, 'start_reason': 'Starting XI', 'end_reason': 'Final Whistle'}]",3869685,France
4,3379,Alphonse Areola,NaN,23,France,[],[],3869685,France


In [13]:
lineups_df.info()
lineups_df.shape # (147 Rows,  9 Column)

<class 'pandas.DataFrame'>
RangeIndex: 147 entries, 0 to 146
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   player_id        147 non-null    int64 
 1   player_name      147 non-null    str   
 2   player_nickname  52 non-null     object
 3   jersey_number    147 non-null    int64 
 4   country          147 non-null    str   
 5   cards            147 non-null    object
 6   positions        147 non-null    object
 7   match_id         147 non-null    int64 
 8   team             147 non-null    str   
dtypes: int64(3), object(3), str(3)
memory usage: 10.5+ KB


(147, 9)

In [14]:
[lineups_df.columns]

[Index(['player_id', 'player_name', 'player_nickname', 'jersey_number',
        'country', 'cards', 'positions', 'match_id', 'team'],
       dtype='str')]

In [15]:
lineups_df.isna().sum()
# For player nickname, the best way is to keep both columns 
# and create a player_display_name field that displays the nickname if available, and if not displays the name

player_id           0
player_name         0
player_nickname    95
jersey_number       0
country             0
cards               0
positions           0
match_id            0
team                0
dtype: int64

In [27]:
lineups_df['cards'][0]

[{'time': '86:24',
  'card_type': 'Yellow Card',
  'reason': 'Foul Committed',
  'period': 2}]

**Decision** <br>
Separating the ```cards``` from the lineups dataset, link it using player_id:<br>

table_name: events<br>



## Events df

In [28]:
events_df.head()

,50_50,bad_behaviour_card,ball_receipt_outcome,ball_recovery_offensive,ball_recovery_recovery_failure,block_deflection,block_offensive,carry_end_location,clearance_aerial_won,clearance_body_part,clearance_head,clearance_left_foot,clearance_other,clearance_right_foot,counterpress,dribble_nutmeg,dribble_outcome,dribble_overrun,duel_outcome,duel_type,duration,foul_committed_advantage,foul_committed_card,foul_committed_offensive,foul_committed_penalty,foul_committed_type,foul_won_advantage,foul_won_defensive,foul_won_penalty,goalkeeper_body_part,goalkeeper_end_location,goalkeeper_outcome,goalkeeper_position,goalkeeper_technique,goalkeeper_type,id,index,interception_outcome,location,match_id,minute,off_camera,out,pass_aerial_won,pass_angle,pass_assisted_shot_id,pass_body_part,pass_cross,pass_deflected,pass_end_location,pass_goal_assist,pass_height,pass_inswinging,pass_length,pass_outcome,pass_outswinging,pass_recipient,pass_recipient_id,pass_shot_assist,pass_switch,pass_technique,pass_through_ball,pass_type,period,play_pattern,player,player_id,position,possession,possession_team,possession_team_id,related_events,second,shot_aerial_won,shot_body_part,shot_end_location,shot_first_time,shot_freeze_frame,shot_key_pass_id,shot_one_on_one,shot_outcome,shot_statsbomb_xg,shot_technique,shot_type,substitution_outcome,substitution_outcome_id,substitution_replacement,substitution_replacement_id,tactics,team,team_id,timestamp,type,under_pressure,injury_stoppage_in_chain,miscontrol_aerial_won,pass_cut_back,pass_miscommunication,pass_no_touch,block_save_block,dribble_no_touch,goalkeeper_shot_saved_to_post,shot_open_goal,shot_saved_to_post
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0584ee21-e3dd-4d9f-95a0-5b5e84be25c3,1,NaN,NaN,3869685,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,Regular Play,NaN,NaN,NaN,1,Argentina,779,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"{'formation': 433, 'lineup': [{'player': {'id': 6909, 'name': 'Damián Emiliano Martínez'}, 'position': {'id': 1, 'name': 'Goalkeeper'}, 'jersey_number': 23}, {'player': {'id': 29201, 'name': 'Nahuel Molina Lucero'}, 'position': {'id': 2, 'name': 'Right Back'}, 'jersey_number': 26}, {'player': {'id': 20572, 'name': 'Cristian Gabriel Romero'}, 'position': {'id': 3, 'name': 'Right Center Back'}, 'jersey_number': 13}, {'player': {'id': 3090, 'name': 'Nicolás Hernán Otamendi'}, 'position': {'id': 5, 'name': 'Left Center Back'}, 'jersey_number': 19}, {'player': {'id': 5507, 'name': 'Nicolás Alejandro Tagliafico'}, 'position': {'id': 6, 'name': 'Left Back'}, 'jersey_number': 3}, {'player': {'id': 38718, 'name': 'Enzo Fernandez'}, 'position': {'id': 10, 'name': 'Center Defensive Midfield'}, 'jersey_number': 24}, {'player': {'id': 7797, 'name': 'Rodrigo Javier De Paul'}, 'position': {'id': 13, 'name': 'Right Center Midfield'}, 'jersey_number': 7}, {'player': {'id': 27886, 'name': 'Alexis Mac Allister'}, 'position': {'id': 15, 'name': 'Left Center Midfield'}, 'jersey_number': 20}, {'player': {'id': 5503, 'name': 'Lionel Andrés Messi Cuccittini'}, 'position': {'id': 17, 'name': 'Right Wing'}, 'jersey_number': 10}, {'player': {'id': 2995, 'name': 'Ángel Fabián Di María Hernández'}, 'position': {'id': 21, 'name': 'Left Wing'}, 'jersey_number': 11}, {'player': {'id': 29560, 'name': 'Julián Álvarez'}, 'position': {'id': 23, 'name': 'Center Forward'}, 'jersey_number': 9}]}",Argentina,779,00:00:00.000,Starting XI,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,b32679f8-942e-4122-96a2-015caf75e628,2,NaN,NaN,3869685,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,Regular Play,NaN,NaN,NaN,1,Argentina,779,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N

In [34]:
events_df.shape # (11764 rows , 104 columns)

(11764, 104)

In [31]:
events_df.isna().sum()

50_50                             11746
bad_behaviour_card                11760
ball_receipt_outcome              11467
ball_recovery_offensive           11762
ball_recovery_recovery_failure    11740
                                  ...  
block_save_block                  11762
dribble_no_touch                  11763
goalkeeper_shot_saved_to_post     11763
shot_open_goal                    11763
shot_saved_to_post                11763
Length: 104, dtype: int64

In [41]:
[events_df.columns]

[Index(['50_50', 'bad_behaviour_card', 'ball_receipt_outcome',
        'ball_recovery_offensive', 'ball_recovery_recovery_failure',
        'block_deflection', 'block_offensive', 'carry_end_location',
        'clearance_aerial_won', 'clearance_body_part', 'clearance_head',
        'clearance_left_foot', 'clearance_other', 'clearance_right_foot',
        'counterpress', 'dribble_nutmeg', 'dribble_outcome', 'dribble_overrun',
        'duel_outcome', 'duel_type', 'duration', 'foul_committed_advantage',
        'foul_committed_card', 'foul_committed_offensive',
        'foul_committed_penalty', 'foul_committed_type', 'foul_won_advantage',
        'foul_won_defensive', 'foul_won_penalty', 'goalkeeper_body_part',
        'goalkeeper_end_location', 'goalkeeper_outcome', 'goalkeeper_position',
        'goalkeeper_technique', 'goalkeeper_type', 'id', 'index',
        'interception_outcome', 'location', 'match_id', 'minute', 'off_camera',
        'out', 'pass_aerial_won', 'pass_angle', 'pass_ass